### DISCUSSION TITLE
PostgreSQL - RUN  Timed Driver script  -  "Error in virtual user extra characters after close-brace"

### DISCUSSION THREAD

**Discussion by nomade33** [TPC-Council/HammerDB#367] [Q&A] [answered]
Hi there,

I am a databases administrator.
I don't speak and write English well, please forgive me.

I just installed Hammerdb 4.4 on Redhat 8.x x64. I am testing Hammerdb TPROC-C with a postgresql 14 database.
I manage to perform simple tests (test driver script checked) by following your documentation and I get a result in the "transaction counter" tab

On the other hand, for a second test with the option "Timed driver script", I have errors in output and the status is in failure for each of the V.U.

The Error is: "error in virtual user #: extra characters after close-brace"

I reloaded the script can be corrupted and renew an execution and always the same errors. 
Can you guide me to establish a diagnosis of this problem?

Bests regards,
Stephan

**Comment by sm-shaw** [comment, marked as answer]
Hi,
Can you provide a bit more detail around the environment?

You are using the GUI instead of the Command Line (CLI) - is this correct? 

If using the GUI is the system you are running HammerDB on remote or local?  If remote is the GUI refreshing slowly? especially on X-Windows (running a display over a slow network) this can be an issue and potentially cause "extra characters after close-brace" to resolve this you should use a GUI solution such as VNC for remote displays and/or moved to fixed rather than scaling graphics https://www.hammerdb.com/blog/uncategorized/hammerdb-v4-0-new-features-pt2-scalable-uhd-graphics/ in the HammerDB configuration. 

Alternatively, you can use the CLI if GUI displays are an issue - I will follow up with a couple of examples. 

Also just for testing make sure that usernames and passwords don't have uncommented special characters such as '#' that can comment out the rest of the line. These can be used but might need solutions such as https://github.com/TPC-Council/HammerDB/discussions/349 to ensure that HammerDB escapes the special characters correctly.

**Reply by sm-shaw** [reply]
Here is a scripted example for the CLI with PostgreSQL on Linux.
So if you copy the examples into scripts pgbuild.tcl and pgrun.tcl in your HammerDB directory, then run:

`./hammerdbcli auto pgbuild.tcl`

to build the schema, note this builds 1000 warehouses with 64 Virtual Users, please adjust accordingly for your CPU and disk capacity.

```
$ more pgbuild.tcl
dbset db pg
dbset bm TPC-C
vuset logtotemp 0
diset tpcc pg_superuser postgres
diset tpcc pg_superuserpass postgres
diset tpcc pg_count_ware 1000
diset tpcc pg_num_vu 64
buildschema
waittocomplete
quit
``` 

To run the test do: 

`./hammerdbcli auto pgrun.tcl`

with the script below:

This example runs 27 separate tests with an increasing load and records a unique logfile for each in your defined temp directory. Again, adjust accordingly for your system. 
  
```
$ more pgrun.tcl
dbset db pg
dbset bm TPC-C
diset tpcc pg_superuser postgres
diset tpcc pg_defaultdbase postgres
diset tpcc pg_driver timed
diset tpcc pg_rampup 2
diset tpcc pg_duration 5
vuset logtotemp 1
vuset unique 1
loadscript
foreach z {1 2 4 8 12 16 20 24 28 32 36 40 44 48 52 56 60 64 68 72 76 80 84 88 92 96 100 } {
puts "$z VU test"
vuset vu $z
vucreate
vurun
runtimer 480
vudestroy
}
puts "Postgres Test Complete"
```
An example result will be posted to #356 in due course, as these are the scripts being used to test that pull request with PostgreSQL 14.1

**Reply by nomade33** [reply]
Hi,
Effectively, i am using with  the GUI instead of the Command Line (CLI). It is correct.
I am  using the GUI. I am running HammerDB on remote linux server redhat 8.x with postgreSQL client 13 and it connects to an other remote PostgreSQL-14 database on linux Server.
On remote, the GUI refreshing slowly on X-Windows (running a display over a slow network) .

I will familiarize myself with cli and try using CLI and your examples.

Thanks.

Bests regards,
Stephan

**Reply by sm-shaw** [reply]
Hi,
Yes, X-Windows has never performed well over long-distance, and scaling SVG graphics compound this problem. (This is an X-Windows rather than HammerDB issue). 
Firstly, switch off SVG by setting fixed scaling in generic.xml. 

```
<theme>
<scaling>fixed</scaling>
<scaletheme>auto</scaletheme>
<pixelsperpoint>auto</pixelsperpoint>
</theme>
```
Nevertheless, if you want a GUI then over long-distance then VNC is absolutely the solution you should use rather than use X-Windows directly, or as above use the CLI instead.
